### PR TITLE
AO3-4873 Implements strong parameters for Tag Set Nominations

### PR DESCRIPTION
**Pull request by davidstump**
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added tests for any changed functionality? (N/A)
* [x] Have you added the JIRA issue number as the *first* thing in your pull request title (eg: `AO3-1234 Fix thing`)
* [x] Have you updated the JIRA issue with the information below?

## Issue

https://otwarchive.atlassian.net/browse/AO3-4873

## Purpose

This PR adds strong parameters to the Tag Set Nominations controller and includes ForbiddenAttributes protection to the relevant TagSetNomination model.

## Testing

1. Log in
2. Browse > Tags > Tag Sets > New Tag Set
3. Fill in Title
4. Check "Currently taking nominations?"
5. Set fandom, character, relationship, and freeform nomination limits to desired numbers
6. Submit
7. Follow "Nominate" link
8. Fill in the nominations form
9. Submit
10. Follow "Edit"
11. Change some things on the form
12. Submit

## References

No other references.

## Credit

Littlelines - David Stump

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

(he/him)

### CODE REVIEW

File: spec/controllers/tag_set_nominations_controller_spec.rb [hunk 1]
```diff
@@ -906,8 +906,8 @@ def add_freeform_nominations(num)
             fake_login_known_user(random_user)
           end
 
-          it 'redirects and returns an error message' do
-            put :update, id: tag_set_nomination.id, tag_set_id: owned_tag_set.id, tag_set_nomination: {}
+          it 'redirects and returns an error message', focus: true do
```
Comment (sarken) [inline spec/controllers/tag_set_nominations_controller_spec.rb]: Will you need this later? If so, no worries! If not, we don't really filter our tests, so could you remove it?

Comment (davidstump) [inline spec/controllers/tag_set_nominations_controller_spec.rb]: Ugh - sorry about that. Thanks for the catch.

File: app/controllers/tag_set_nominations_controller.rb [hunk 2]
```diff
@@ -1,22 +1,22 @@
...
   before_filter :users_only
   before_filter :load_tag_set, :except => [ :index ]
   before_filter :check_pseud_ownership, :only => [:create, :update]
   before_filter :load_nomination, :only => [:show, :edit, :update, :destroy]
   before_filter :set_limit, :only => [:new, :edit, :show, :create, :update, :review]
 
   def check_pseud_ownership
-    if params[:tag_set_nomination][:pseud_id]
-      pseud = Pseud.find(params[:tag_set_nomination][:pseud_id])
+    if tag_set_nomination_params[:pseud_id]
```
Comment (davidstump) [inline app/controllers/tag_set_nominations_controller.rb]: I am happy to revert if that's the style you prefer for your repo. I have generally used the strong param methods anywhere that references the main model attributes (such as tag_set_nomination) as a general rule. With the new method, you get a specialized required attribute error vs a slightly more vague error about attempting to call `[]` off `nil` if a parameter was not sent properly. Just let me know which style you prefer - and happy to change.

Comment (sarken) [inline app/controllers/tag_set_nominations_controller.rb]: More specific error messages sound good, so let's use this one. Thanks for the explanation!

**Comment by tickinginstant** [comment]
Sorry for butting in, but I believe the set of whitelisted params is incomplete. I think the code should be:
```
    params.require(:tag_set_nomination).permit(
      :pseud_id,
      :owned_tag_set_id,
      fandom_nominations_attributes: [
        :id,
        :tagname,
        character_nominations_attributes: [
          :id, :tagname, :from_fandom_nomination
        ],
        relationship_nominations_attributes: [
          :id, :tagname, :from_fandom_nomination
        ],
      ],
      freeform_nominations_attributes: [
        :id, :tagname
      ],
      character_nominations_attributes: [
        :id, :tagname, :parent_tagname
      ],
      relationship_nominations_attributes: [
        :id, :tagname, :parent_tagname
      ]
    )
```
I believe the extra `character_nominations_attributes` and `relationship_nominations_attributes` are needed because tag sets that don't allow fandom nominations behave somewhat differently from tag sets that do. In particular, the CharacterNominations and RelationshipNominations aren't nested in the FandomNominations, because there are no FandomNominations. And they have an additional `parent_tagname` field for entering the fandom associated with each character and/or relationship.

(Unfortunately, this doesn't seem to be covered by test cases. I'm guessing that's why this case slipped through the cracks. There's [one cucumber test for a tagset with 0 fandom nominations and 3 character nominations](https://github.com/otwcode/otwarchive/blob/master/features/tags_and_wrangling/tag_set.feature#L62), but while it does open the nomination form, it doesn't actually submit any nominations. There are a few RSpec tests that set `fandom_nomination_limit` to 0, but [one group of tests is for the `index` action](https://github.com/otwcode/otwarchive/blob/master/spec/controllers/tag_set_nominations_controller_spec.rb#L296) and [the other group of tests is for the `new` action](https://github.com/otwcode/otwarchive/blob/master/spec/controllers/tag_set_nominations_controller_spec.rb#L568) -- no tests with 0 fandoms for `create` or `update`.)

**Comment by sarken** [comment]
Thanks for catching that, @tickinginstant! Would you by any chance be able to write a test for that? If not, no worries -- I will add one to #2788

**Comment by tickinginstant** [comment]
Sure, I can write a test. Should I submit it with the same issue number as #2788?

**Comment by sarken** [comment]
Thank you so much! I've made a shiny new issue for you here: https://otwarchive.atlassian.net/browse/AO3-4908 (I'll update the description in a minute)

**Comment by sarken** [comment]
I ran this strong parameters code through Travis with @tickinginstant's new tests (which are now merged to master), and it did indeed fail: https://travis-ci.org/sarken/otwarchive/builds/209273689

When I tried it with tickinginstant's code, it passed: https://travis-ci.org/sarken/otwarchive/builds/209274530

(If you want, you can pull from my branch -- https://github.com/sarken/otwarchive/tree/AO3-4873 -- but that might be more trouble than pulling from otwarchive master and then copy & pasting tickinginstant's code.)

**Comment by sarken** [comment]
Hi, David! Would you have a few minutes to resolve the merge conflict on this so we can merge it? (It looks like it's a few lines of code that have been deleted.)

**Comment by sarken** [comment]
Oh, sorry -- for some reason, I thought tickinginstant's comments had already been actioned and this just needed merge conflicts resolved. I will just... quietly shuffle this out of this release, and you can address it whenever you have time!

**Comment by davidstump** [comment]
@sarken No problem - I went ahead and added the additional params and pushed. Thanks!

**Pull request merged**